targets/TARGET_STM/qspi_api.c

```diff
@@ -0,0 +1,245 @@
+            st_command->AddressMode = QSPI_ADDRESS_1_LINE;
+            break;
+        case QSPI_CFG_BUS_DUAL:
+            st_command->AddressMode = QSPI_ADDRESS_2_LINES;
+            break;
+        case QSPI_CFG_BUS_QUAD:
+            st_command->AddressMode = QSPI_ADDRESS_4_LINES;
+            break;
+        default:
+            st_command->AddressMode = QSPI_ADDRESS_1_LINE;
+            break;
+    }
+
+    if (command->address.size == QSPI_CFG_ADDR_SIZE_NONE) {
+        st_command->AddressMode = QSPI_ADDRESS_NONE;
+    }
+
+    switch (command->alt.bus_width) {
+        case QSPI_CFG_BUS_SINGLE:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_1_LINE;
+            break;
+        case QSPI_CFG_BUS_DUAL:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_2_LINES;
+            break;
+        case QSPI_CFG_BUS_QUAD:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_4_LINES;
+            break;
+        default:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_1_LINE;
+            break;
+    }
+
+    if (command->alt.size == QSPI_CFG_ALT_SIZE_NONE) {
+        st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_NONE;
+    }
+
+    switch (command->data.bus_width) {
+        case QSPI_CFG_BUS_SINGLE:
+            st_command->DataMode = QSPI_DATA_1_LINE;
+            break;
+        case QSPI_CFG_BUS_DUAL:
+            st_command->DataMode = QSPI_DATA_2_LINES;
+            break;
+        case QSPI_CFG_BUS_QUAD:
+            st_command->DataMode = QSPI_DATA_4_LINES;
+            break;
+        default:
+            st_command->DataMode = QSPI_DATA_1_LINE;
+            break;
+    }
+}
+
+
+qspi_status_t qspi_init(qspi_t *obj, PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName ssel, uint32_t hz, uint8_t mode)
+{
+    // Enable interface clock for QSPI
+     __HAL_RCC_QSPI_CLK_ENABLE();
+
+    // Reset QSPI
+    __HAL_RCC_QSPI_FORCE_RESET();
+    __HAL_RCC_QSPI_RELEASE_RESET();
+
+    // Set default QSPI handle values
+    obj->handle.Init.ClockPrescaler = 1;
+    obj->handle.Init.FifoThreshold = 4;
+    obj->handle.Init.SampleShifting = QSPI_SAMPLE_SHIFTING_HALFCYCLE;
+    obj->handle.Init.FlashSize = POSITION_VAL(QSPI_FLASH_SIZE_DEFAULT) - 1;
+    obj->handle.Init.ChipSelectHighTime = QSPI_CS_HIGH_TIME_2_CYCLE;
+    obj->handle.Init.ClockMode = QSPI_CLOCK_MODE_0;
+    obj->handle.Init.FlashID = QSPI_FLASH_ID_1;
+    obj->handle.Init.DualFlash = QSPI_DUALFLASH_DISABLE;
+
+    obj->handle.Init.ClockMode = mode == 0 ? QSPI_CLOCK_MODE_0 : QSPI_CLOCK_MODE_3;
+
+    obj->handle.Instance = QUADSPI;
+
+    // TODO pinmap here for pins (enable clock)
+    pinmap_pinout(io0, PinMap_QSPI_DATA);
+    pinmap_pinout(io1, PinMap_QSPI_DATA);
+    pinmap_pinout(io2, PinMap_QSPI_DATA);
+    pinmap_pinout(io3, PinMap_QSPI_DATA);
+
+    pinmap_pinout(sclk, PinMap_QSPI_SCLK);
+    pinmap_pinout(ssel, PinMap_QSPI_SSEL);
+
+    if (HAL_QSPI_Init(&obj->handle) != HAL_OK) {
```

Comment:
Question: what about the DMA configuration as done CubeF4 QSPI example ?

Reply:
DMA not at the moment